checker/src/body_visitor.rs

```diff
@@ -2248,4 +2256,53 @@ impl<'analysis, 'compilation, 'tcx, E> BodyVisitor<'analysis, 'compilation, 'tcx
 
         (tag_field_path, tag_field_value)
     }
+
+    /// Attach a tag to implicit tag field paths that are rooted by root_path.
+    /// If v is the value at a tag field path, then it is updated to attach_tag(v).
+    fn propagate_tag_to_implicit_tag_fields<F>(&mut self, root_path: Rc<Path>, attach_tag: F)
+    where
+        F: Fn(Rc<AbstractValue>) -> Rc<AbstractValue>,
+    {
+        // Record visited prefixes of paths. We need this information to avoid handling the same prefix for multiple times.
+        let mut visited_path_prefixes: HashSet<Rc<Path>> = HashSet::new();
+
+        let old_value_map = self.current_environment.value_map.clone();
+
+        for (path, _) in old_value_map
+            .iter()
+            .filter(|(p, _)| p.is_rooted_by(&root_path))
+        {
+            let mut path_prefix = path;
+            while let PathEnum::QualifiedPath { qualifier, .. } = &path_prefix.value {
+                path_prefix = qualifier;
+                if visited_path_prefixes.contains(path_prefix)
+                    || !path_prefix.is_rooted_by(&root_path)
+                {
+                    break;
+                } else {
+                    visited_path_prefixes.insert(path_prefix.clone());
+
+                    let path_prefix_rustc_type = self
+                        .type_visitor
+                        .get_path_rustc_type(path_prefix, self.current_span);
+
+                    if !path_prefix_rustc_type.is_scalar() {
+                        let (tag_field_path, tag_field_value) = self
+                            .extract_tag_field_of_non_scalar_value_at(
+                                path_prefix,
+                                path_prefix_rustc_type,
+                            );
+
+                        // Perform an update when tag_field_path is not tracked in the old environment.
+                        // In this case, tag_field_path represents an implicit tag field.
```

Comment:
It is not clear to me why you attach the tag only if the prefix does not already have a tag field. I would like an explanation and a test case to prove it is not needed.

Reply:
Those existing tag fields are already updated by non_patterned_copy_or_move. Added a comment and a test to clarify this.